collab2/foraging/toolkit/local_windows.py

```diff
@@ -0,0 +1,121 @@
+from itertools import product
+from typing import Any, Callable, List, Optional
+
+import numpy as np
+import pandas as pd
+
+
+def get_grid(
+    grid_size: int = 90,
+    sampling_fraction: float = 1.0,
+    random_seed: int = 0,
+    grid_constraint: Optional[
+        Callable[[pd.DataFrame, pd.DataFrame, Optional[dict]], pd.DataFrame]
+    ] = None,
+    grid_constraint_params: Optional[dict] = None,
+) -> pd.DataFrame:
+    # generate grid of all points
+    mesh = product(range(grid_size), repeat=2)
+    grid = pd.DataFrame(mesh, columns=["x", "y"])
+
+    # only keep accessible points
+    if grid_constraint is not None:
+        grid = grid[grid_constraint(grid["x"], grid["y"], grid_constraint_params)]
+
+    # subsample the grid
+    np.random.seed(random_seed)
+    drop_ind = np.random.choice(grid.index, int(len(grid) * (1 - sampling_fraction)))
+    grid = grid.drop(drop_ind)
+
+    return grid
+
+
+def _generate_local_windows(
+    foragers: List[pd.DataFrame],
+    grid_size: int,
+    num_foragers: int,
+    num_frames: int,
+    window_size: float,
+    sampling_fraction: float = 1.0,
+    random_seed: int = 0,
+    skip_incomplete_frames: bool = False,
+    grid_constraint: Optional[
+        Callable[[pd.DataFrame, pd.DataFrame, Any], pd.DataFrame]
+    ] = None,
+    grid_constraint_params: Optional[dict] = None,
+) -> pd.DataFrame:
+
+    # Note: args grid_size, num_foragers, num_frames are not exposed to users but set to
+    # values inherited from foragers_object by generate_local_windows
+
+    # initialize a common grid
+    grid = get_grid(
+        grid_size=grid_size,
+        sampling_fraction=sampling_fraction,
+        random_seed=random_seed,
+        grid_constraint=grid_constraint,
+        grid_constraint_params=grid_constraint_params,
+    )
+
+    f_present_frames = []
+    for f in range(num_foragers):
+        tracked_idx = foragers[f].loc[:, ["x", "y"]].notna().all(axis=1)
+        f_present_frames.append(foragers[f]["time"].loc[tracked_idx].to_list())
+
+    # identify time points where ALL foragers are present
+    f_present_frames_set = [set(_) for _ in f_present_frames]
+    all_present_frames = set.intersection(*f_present_frames_set)
+
+    # calculate local_windows for each forager
+    local_windows = []
+    for f in range(num_foragers):
+        # initialize local_windows_f to None
+        local_windows_f = [None for _ in range(num_frames)]
+
+        # find frames for which local windows need to be computed
+        if skip_incomplete_frames:
+            compute_frames = all_present_frames
+        else:
+            compute_frames = f_present_frames[f]
+
+        for t in compute_frames:
+            # copy grid
```

Comment:
try using copy.deepcopy() in line 83 and then simplifying this bit, see what happens